From eb5c0aa693733e727817dfc41bfbe1efa2c117c0 Mon Sep 17 00:00:00 2001
From: Haoyang Chen <chenhaoyang@uniontech.com>
Date: Mon, 14 Sep 2020 11:02:19 +0800
Subject: [PATCH 10/51] From 43ea0be463f43e0d2bc6b29603d40ee4996a2e98 Mon Sep
 17 00:00:00 2001 From: Haoyang Chen <chenhaoyang@uniontech.com> Date: Fri, 11
 Sep 2020 17:06:27 +0800 Subject: [PATCH] ntdll: Improve the performance of
 log output.

Logs with line breaks do not need to be copied.

Signed-off-by: Haoyang Chen <chenhaoyang@uniontech.com>
---
 dlls/ntdll/unix/debug.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/dlls/ntdll/unix/debug.c b/dlls/ntdll/unix/debug.c
index d4aad491edc..b902502830c 100644
--- a/dlls/ntdll/unix/debug.c
+++ b/dlls/ntdll/unix/debug.c
@@ -308,9 +308,18 @@ int __cdecl __wine_dbg_output( const char *str )
 
     if (end)
     {
-        ret += append_output( info, str, end + 1 - str );
-        write( 2, info->output, info->out_pos );
-        info->out_pos = 0;
+        if (info->out_pos)
+        {
+            ret += append_output( info, str, end + 1 - str );
+            write( 2, info->output, info->out_pos );
+            info->out_pos = 0;
+        }
+        else
+        {
+            /* output directly, no need cache */
+            ret = end - str + 1;
+            write( 2, str, ret );
+        }
         str = end + 1;
     }
     if (*str) ret += append_output( info, str, strlen( str ));
-- 
2.53.0

